### PR TITLE
Split circleci jobs.

### PR DESCRIPTION
**Pull request by chriseth**
The biggest problem with this is that it still builds some stuff in the "test" jobs.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
+  executor:
     machine:
       image: ubuntu-2004:current
     resource_class: large
+
+jobs:
+  build:
+    executor: executor
     steps:
+      - checkout
```
Comment (leonardoalt) [inline .circleci/config.yml:13]: how can it do checkout without git? or do we assume the image has git?

File: .circleci/config.yml [hunk 2]
```diff
@@ -41,51 +52,121 @@ jobs:
...
-          command: rustup toolchain install nightly-2023-01-03-x86_64-unknown-linux-gnu
-
-      - run:
-          name: Install riscv target
-          command: rustup target add riscv32imac-unknown-none-elf --toolchain nightly-2023-01-03-x86_64-unknown-linux-gnu
-
-      - run:
-          name: Install stdlib
-          command: rustup component add rust-src --toolchain nightly-2023-01-03-x86_64-unknown-linux-gnu
+            - v5-cargo-pr-tests-{{ .Branch }}
```
Comment (leonardoalt) [inline .circleci/config.yml:55]: I'm not sure I agree with this. The cache from `main` can also be useful.

File: .circleci/config.yml [hunk 3]
```diff
@@ -41,51 +52,121 @@ jobs:
...
+            rustup toolchain install 1.72-x86_64-unknown-linux-gnu
+            rustup toolchain install nightly-2023-01-03-x86_64-unknown-linux-gnu
+            rustup target add riscv32imac-unknown-none-elf --toolchain nightly-2023-01-03-x86_64-unknown-linux-gnu
+            rustup component add rust-src --toolchain nightly-2023-01-03-x86_64-unknown-linux-gnu
+
+      - run:
+          name: Determine which tests to run
+          command: |
+              cargo test --all --all-features --profile pr-tests  -- --list 2> /dev/null | grep ': test' | sed -e 's/: test//' | \
+              circleci tests split --split-by=timings | tee tests_to_run.txt
```
Comment (leonardoalt) [inline .circleci/config.yml:109]: is there no better way of doing this?

File: .circleci/config.yml [hunk 4]
```diff
@@ -41,51 +52,121 @@ jobs:
...
           paths:
             - ~/.cargo/registry
             - ~/.cargo/git
             - target
 
+      - persist_to_workspace:
+          root: .
+          paths:
+            - target/pr-tests
+            - pilcom
```
Comment (leonardoalt) [inline .circleci/config.yml:77]: we could also keep `.rustup` and `.cargo` so we don't need to reinstall them

File: .circleci/config.yml (hunk 1 above)
Comment (chriseth) [reply .circleci/config.yml:13]: It's a circleci-specific thing.

File: .circleci/config.yml (hunk 2 above)
Comment (chriseth) [reply .circleci/config.yml:55]: Yes, as a fallback we should use the cache from `main`. The problem is that it should be somewhat recent. I couldn't find a good way to do this yet.

File: .circleci/config.yml (hunk 3 above)
Comment (chriseth) [reply .circleci/config.yml:109]: It's the best way I could find.

File: .circleci/config.yml (hunk 4 above)
Comment (chriseth) [reply .circleci/config.yml:77]: We should use a custom docker image.

File: .circleci/config.yml (hunk 2 above)
Comment (leonardoalt) [reply .circleci/config.yml:55]: I think you can add an extra line with another cache key that is checked if the first one fails. We can actively invalidate it once in a while

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**